### DISCUSSION TITLE
Cannot read properties of undefined (reading 'createContext')

### DISCUSSION THREAD

**Discussion by closertb** [webpack/webpack#15125] [Q&A]
# Bug report
i upgrade my webpack version to Webpack5; output.module is systemJs，it compile success, but the page log the error:

> Cannot read properties of undefined (reading 'createContext')

![image](https://user-images.githubusercontent.com/22979584/148528190-617dd7e5-c7d2-4cbf-a7f3-2d5141f89159.png)

i look the difference between from v4 from v5, i find that:

webpack5：

![image](https://user-images.githubusercontent.com/22979584/148527776-65c27f4b-c6bf-49a6-9252-8ca683d6f9a3.png)

webpack4:

![image](https://user-images.githubusercontent.com/22979584/148528059-3a53949a-37ad-47cc-95d4-076f73bf7f12.png)

webpack4 use __webpack_require__.n to compatible the 'default' exports

**What is the current behavior?**

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

the page run ok

**Other relevant information:**
webpack version:  5.65.0
Node.js version: v14.18.0
Operating System:  Mac
Additional tools:

**Comment by alexander-akait** [comment]
Please create reproducible test repo

**Reply by closertb** [reply]
> Please create reproducible test repo

https://github.com/closertb/webpackDemo

clone it, and look the readme

**Reply by alexander-akait** [reply]
You have mixed webpack versions (i.e. webpack v4 and webpack v5, run `npm ls webpack`, it can be big problem), some plugins are outdated and don't work with webpack v5, please fix these problems firstly, some loaders too, for example -  `css-loader`, and I can't run your repo due:
> npm ERR! 404 Not Found - GET https://registry.npmjs.org/@ali%2fbuilder-saas-utils - Not found
Looks like you use custom register, so I can't download this package.

I tried reproduce the problem locally and all works fine, I think the problem somewhere in old plugins or wrong configuration

**Reply by closertb** [reply]
> You have mixed webpack versions (i.e. webpack v4 and webpack v5, run `npm ls webpack`, it can be big problem), some plugins are outdated and don't work with webpack v5, please fix these problems firstly, some loaders too, for example - `css-loader`, and I can't run your repo due:
> 
> > npm ERR! 404 Not Found - GET https://registry.npmjs.org/@ali%2fbuilder-saas-utils - Not found
> > Looks like you use custom register, so I can't download this package.
> 
> I tried reproduce the problem locally and all works fine, I think the problem somewhere in old plugins or wrong configuration

follow your advice, I update  the demo...and I found the difference from webpack5 from 4

![image](https://user-images.githubusercontent.com/22979584/148968910-e6b48a5f-ea7c-45c6-a262-cddd3e0fef65.png)

![image](https://user-images.githubusercontent.com/22979584/148968980-c4755bd9-b841-4645-a11a-f1395bcc4c4f.png)


webpack5 system module treat all the external to esmodule...but the react and react dom are not, they just offer the commonjs module, no default output

**Reply by alexander-akait** [reply]
What is `external` configuration?

**Reply by closertb** [reply]
> What is `external` configuration?

```js
config.externals({
  react: 'React',
  'react-dom': 'ReactDOM',
});
```

and i register it at the head tag,  please download the demo

**Reply by alexander-akait** [reply]
You can try:
```
config.externals({
  react: 'commonjs-module React',
  'react-dom': 'commonjs-module ReactDOM',
});
```

**Reply by closertb** [reply]
> 

I had, but did not work

**Reply by closertb** [reply]
> You can try:
> 
> ```
> config.externals({
>   react: 'commonjs-module React',
>   'react-dom': 'commonjs-module ReactDOM',
> });
> ```

it throw error: require is not defined....did i lost some config？ the require is a commonjs syntax, not  exsit in the browser

**Reply by alexander-akait** [reply]
What is `target`? Browser or node?

**Reply by closertb** [reply]
> 

browser....you can git pull my demo:  https://github.com/closertb/webpackDemo,  and try to run it.... see the readme

**Reply by closertb** [reply]
> What is `target`? Browser or node?

did you have a look ?

**Reply by alexander-akait** [reply]
Firstly - You have mixed webpack version, please migrate from `optimize-css-assets-webpack-plugin` on https://github.com/webpack-contrib/css-minimizer-webpack-plugin and update `thread-loader`

Using your repo I can't get your error, my output:
```
res {} undefined
(index):21 e undefined
(index):22 typeof exports: undefined typeof module: undefined typeof define: undefined
(index):36 res {} undefined
log.js:24 [HMR] Waiting for update signal from WDS...
index.jsx:6 use--------
index.js:551 [webpack-dev-server] Hot Module Replacement enabled.
index.js:551 [webpack-dev-server] Live Reloading enabled.
```

And no errors, maybe I missed something?

**Comment by closertb** [comment]
Browser.....

you can git pull my demo:  https://github.com/closertb/webpackDemo

**Reply by herrera-ignacio** [reply]
Any solution?

**Reply by vankop** [reply]
could you please reduce you build config to single `webpack.config.js`?

**Reply by closertb** [reply]
> could you please reduce you build config to single `webpack.config.js`?

you can open my demo above:  https://github.com/closertb/webpackDemo

flow the steps of readme
